Increase test coverage for replacing ops based on gas price

hazim-j opened the pull request:
Propose a fix for the two test cases in `tests/bundle/test_bundle.py`.

1. `test_bundle_replace_by_fee` currently doesn't take into account an equally increased `maxFeePerGas`.
2. `test_bundle` sends multiple ops to the bundler from an unstaked account.

As is, both test cases don't comply with the sanity check for pending userOps in the mempool: https://github.com/eth-infinitism/account-abstraction/blob/c2a541bb8c29cbb583c5ea1f3eefdd99be40654c/eip/EIPS/eip-4337.md?plain=1#L276

shahafn commented:
Hi @hazim-j thank you for the PR!
1. Yes you're right. Some tests are missing and that's one of them.
2. Yes the eip currently isn't updated, we're adding that to the eip and it will be published soon.

shahafn commented on tests/bundle/test_bundle.py:
```diff
@@ -1,31 +1,42 @@
         nonce="0x1",
         callData=callData,
-        maxPriorityFeePerGas=hex(10**9),
+        maxPriorityFeePerGas=lowerFeePerGas,
```
It's probably best to separate it to two different tests, one for maxFeePerGas and another for maxPriorityFeePerGas

hazim-j commented:
@shahafn finally got around to adding the two tests.

1. `test_bundle_replace_with_only_priority_fee_change`: changing `maxPriorityFeePerGas` without equally changing `maxFeePerGas` should fail regardless of whether new priority fee is higher or lower.
2. `test_bundle_replace_with_equally_increasing_max_fee`: If `maxFeePerGas` is equally increased, then it should replace as long as priority fee is higher.

hazim-j commented:
@shahafn Done. This PR should be up to date now.

hazim-j commented:
@shahafn I fixed the lint issue and all tests are passing except for `test_bundle_replace_with_only_priority_fee_change` which I added in this PR. This would be expected if the reference implementation has not yet implemented that logic.

shahafn commented:
Hi @hazim-j after looking again at what geth does, I believe that these checks should change to take into account only the effective gas price. That is, a transaction should be replaced only if both maxFeePerGas and maxPriorityFeePerGas were increased by at least a factor (defaults to 10%). 
See here
https://github.com/ethereum/go-ethereum/blob/06632da2bb367664b5e1417c55c0d3a2a2bf5166/core/txpool/list.go#L282

https://github.com/ethereum/go-ethereum/blob/793f0f9ec860f6f51e0cec943a268c10863097c7/core/txpool/txpool.go#L179

Geth is only replacing transactions if they were increased by priceBump percentage, which defaults to 10.

hazim-j commented:
> Geth is only replacing transactions if they were increased by priceBump percentage, which defaults to 10.

@shahafn I'll look into this and update the PR sometime within the next week. If this is the case, should the EIP also be updated for more context?

shahafn commented:
I think that the because this is an inherited behavior from Ethereum nodes, it is implied in the EIP. I'll check though to make sure. Thanks!

shahafn commented:
Hey @hazim-j following up on the PR fixes, are you able to introduce them this week?

hazim-j commented:
@shahafn I had a look at the geth logic and you're right. In order to be replaced, both `maxFeePerGas` and `maxPriorityFeePerGas` must increase by at least 10%. Given this logic, I propose adding the following test cases:

- `test_bundle_replace_with_only_priority_fee_bump`
- `test_bundle_replace_with_only_max_fee_bump`
- `test_bundle_replace_with_fee_reduction`
- `test_bundle_replace_with_fee_bump_below_threshold`
- `test_bundle_replace_with_fee_bump_at_threshold`
- `test_bundle_replace_with_fee_bump_above_threshold`

I also suspect that the one failing test in the build is due to this line in the reference bundler: https://github.com/eth-infinitism/bundler/blob/a27f1b57a6d22b24c8d524c6569579f5e78522a7/packages/bundler/src/modules/MempoolManager.ts#L81

At the threshold value, the op should still be accepted. So this line should instead be `newGas >= oldGas * 1.1`. Also the case `test_bundle_replace_with_fee_bump_above_threshold` seems to pass which makes it likely a comparison error.

drortirosh commented on tests/bundle/test_bundle.py:
```diff
@@ -11,38 +11,239 @@
+
+@pytest.mark.parametrize("mode", ["manual"], ids=[""])
+@pytest.mark.usefixtures("clear_state", "set_bundling_mode")
+def test_bundle_replace_with_only_priority_fee_bump(w3):
```
These tests seems ok, but it is very hard to read and verify, as it is a copy-and-paste of a lot of information (with minor modifications)

Most tests are basically:
`test_replace_fee(bump_prio, bump_max, outcome)`
where "outcome" is a function that gets 2 ops, expects the 1st to succeed and the 2nd is either replace or fail.

the above test function can then be parameterized by adding before it:
```
@pytest.mark.parametrize("bump_prio,bump_max,outcome",[
  [0,0,fail_replace],
  [10,0,fail_replace],
...
  [10,10,succeed_replace]
])
```

hazim-j replied:
Agree that parameterising the test would be a lot better for readability. I will refactor in the next few days.

hazim-j commented:
@drortirosh I've refactored the test cases to be parameterized (using a similar pattern to `test_storage_rules.py`). It should be much more readable now. Let me know if there are any issues.

shahafn commented:
Hi @hazim-j , I tested your PR with my fix in https://github.com/eth-infinitism/bundler/pull/68 and it passes. Once we introduce this fix to the bundler I'll merge this PR as well.
